### PR TITLE
TF-3004 [SEARCH] Open close advanced search looses data

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#3004

## Resolved

https://github.com/user-attachments/assets/0205b4d5-6f4c-458c-8c82-4060ee1bd0e8

### CODE REVIEW

**Comment by dab246** [comment]
- [ ] Add SortOrder to search filter

**Comment by dab246** [comment]
> please rebase it @dab246

done

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart [hunk 1]
```diff
@@ -80,87 +79,128 @@ class AdvancedFilterController extends BaseController {
...
+      emailReceiveTimeTypeOption: emailReceiveTimeTypeOption,
+      hasAttachmentOption: hasAttachmentOption,
+      beforeOption: beforeOption,
+      startDateOption: startDateOption,
+      endDateOption: endDateOption,
+      positionOption: positionOption,
+      sortOrderTypeOption: sortOrderTypeOption);
+  }
+
+  void _updateDataToMemorySearchFilter() {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:119]:
```suggestion
  void _synchronizeSearchFilter() {
```

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart [hunk 2]
```diff
@@ -496,6 +551,20 @@ class AdvancedFilterController extends BaseController {
     }
   }
 
+  void _handleSearchEmailByFromFieldsAction(EmailAddress emailAddress) {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:554]:
```suggestion
  void _handleQuickSearchEmailByFromAction(EmailAddress emailAddress) {
```

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart [hunk 3]
```diff
@@ -399,23 +462,12 @@ class AdvancedFilterController extends BaseController {
           );
         } else if (action is ClearDateRangeToAdvancedSearch) {
           _updateDateRangeTime(action.receiveTime);
-        } else if (action is StartSearchEmailBySearchFilterAction) {
-          _handleSearchEmailBySearchFilterAction(action.searchFilter);
+        } else if (action is StartSearchEmailAction) {
+          _handleStartSearchEmailAction();
         } else if (action is SearchEmailByFromFieldsAction) {
```
[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:467]: And change the name of method `searchEmailByFromFields` in `MailboxDashBoardController` to `quickSearchEmailByFrom`

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart [hunk 4]
```diff
@@ -183,43 +223,48 @@ class AdvancedFilterController extends BaseController {
...
-        mailBoxFilterInputController.text = AppLocalizations.of(currentContext!).allFolders;
-      } else {
-        mailBoxFilterInputController.text = StringConvert.writeNullToEmpty(
-          searchEmailFilter.mailbox?.getDisplayName(currentContext!)
-        );
-      }
-    }
-    hasAttachment.value = searchEmailFilter.hasAttachment;
-    if (searchEmailFilter.from.isEmpty) {
+  void initSearchFilterField(BuildContext? context) {
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:226]: Please write unit test for this method

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart [hunk 5]
```diff
@@ -440,38 +492,41 @@ class AdvancedFilterController extends BaseController {
...
+  void onHasAttachmentCheckboxChanged(bool? isChecked) {
+    hasAttachment.value = isChecked ?? false;
+    _updateMemorySearchFilter(hasAttachmentOption: Some(hasAttachment.value));
   }
 
-  void _handleSearchEmailBySearchFilterAction(QuickSearchFilter searchFilter) {
-    switch(searchFilter) {
-      case QuickSearchFilter.from:
-        _updateFromField();
+  void onTextChanged(AdvancedSearchFilterField filterField, String value) {
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:513]: Please write unit test for this method

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:226]: What is the purpose?

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 5 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:513]: What is the purpose?

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:119]: done

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 2 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:554]: done

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:467]: done

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 4 above)
Comment (tddang-linagora) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:226]: To ensure its functionality

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 5 above)
Comment (tddang-linagora) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:513]: idem

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 4 above)
Comment (tddang-linagora) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:226]: - Confirm controller's values after `initSearchFilterField` is called when
  - [ ] _memorySearchFilter.from is empty
  - [ ] _memorySearchFilter.from is not empty
  - [ ] _memorySearchFilter.to is empty
  - [ ] _memorySearchFilter.to is not empty
  - [ ] _memorySearchFilter.mailbox == null
  - [ ] _memorySearchFilter.mailbox != null

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 5 above)
Comment (tddang-linagora) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:513]: - Confirming _updateMemorySearchFilter is called with expected values when
  - [ ] `filterField` is `AdvancedSearchFilterField.subject`
  - [ ] `filterField` is `AdvancedSearchFilterField.hasKeyword`
  - [ ] `filterField` is `AdvancedSearchFilterField.notKeyword`
- [ ] Confirming nothing change when `filterField` is none of the above

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart [hunk 6]
```diff
@@ -80,87 +79,128 @@ class AdvancedFilterController extends BaseController {
...
+
+    _destinationMailboxSelected = destinationMailbox;
+    final mailboxName = context.mounted
+      ? _destinationMailboxSelected?.getDisplayName(context)
+      : _destinationMailboxSelected?.name?.name;
+    mailBoxFilterInputController.text = StringConvert.writeNullToEmpty(mailboxName);
+    _updateMemorySearchFilter(mailboxOption: optionOf(_destinationMailboxSelected));
   }
 
   void applyAdvancedSearchFilter() {
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:198]: done

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:226]: done

File: lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart (hunk 5 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/advanced_filter_controller.dart:513]: done

**Review by tddang-linagora — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**